master/buildbot/test/unit/test_reporter_zulip.py

```diff
@@ -0,0 +1,183 @@
+                "event": "new",
+                "buildid": 20,
+                "buildername": "Builder0",
+                "url": "http://localhost:8080/#builders/79/builds/0",
+                "project": "testProject",
+                "timestamp": 1554161923
+            })
+        self.sp.buildStarted(('build', 20, 'new'), build)
+
+    @defer.inlineCallbacks
+    def test_build_finished(self):
+        yield self.setupZulipStatusPush(stream="xyz")
+        build = yield self.setupBuildResults()
+        build["complete_at"] = datetime.datetime(2019, 4, 1, 23, 38, 43, 154354, tzinfo=tzutc())
+        self._http.expect(
+            'post',
+            '/api/v1/external/buildbot?api_key=123&stream=xyz',
+            json={
+                "event": "finished",
+                "buildid": 20,
+                "buildername": "Builder0",
+                "url": "http://localhost:8080/#builders/79/builds/0",
+                "project": "testProject",
+                "timestamp": 1554161923,
+                "results": 0
+            })
+        self.sp.buildFinished(('build', 20, 'finished'), build)
+
+    @defer.inlineCallbacks
+    def test_stream_none(self):
+        yield self.setupZulipStatusPush(stream=None)
+        build = yield self.setupBuildResults()
+        build["complete_at"] = datetime.datetime(2019, 4, 1, 23, 38, 43, 154354, tzinfo=tzutc())
+        self._http.expect(
+            'post',
+            '/api/v1/external/buildbot?api_key=123',
+            json={
+                "event": "finished",
+                "buildid": 20,
+                "buildername": "Builder0",
+                "url": "http://localhost:8080/#builders/79/builds/0",
+                "project": "testProject",
+                "timestamp": 1554161923,
+                "results": 0
+            })
+        self.sp.buildFinished(('build', 20, 'finished'), build)
+
+    def test_endpoint_string(self):
+        with self.assertRaisesConfigError(
+                "Endpoint must be a string"):
+            ZulipStatusPush(endpoint=1234, token="abcd")
+
+    def test_token_string(self):
+        with self.assertRaisesConfigError(
+                "Token must be a string"):
+            ZulipStatusPush(endpoint="http://example.com", token=1234)
+
+    @defer.inlineCallbacks
+    def test_invalid_json_data(self):
+        yield self.setupZulipStatusPush(stream="xyz")
+        build = yield self.setupBuildResults()
+        build["started_at"] = datetime.datetime(2019, 4, 1, 23, 38, 43, 154354, tzinfo=tzutc())
+        self._http.expect(
+            'post',
+            '/api/v1/external/buildbot?api_key=123&stream=xyz',
+            json={
+                "event": "new",
+                "buildid": 20,
+                "buildername": "Builder0",
+                "url": "http://localhost:8080/#builders/79/builds/0",
+                "project": "testProject",
+                "timestamp": 1554161923
+            }, code=500)
+        self.setUpLogging()
```

Comment:
`setUpLogging` should go to the `setUp()` method.

Reply:
Do we really need to setup logging for each test? Only a couple of the test use it so won't it be better to just set it up inside those tests?

Reply:
Agreed, now I see that a lot of tests already do that.